src/constraint.rs

```diff
@@ -73,6 +78,79 @@ where
     Box::new(Impossible { constraint })
 }
 
+pub fn implies<Var, Value>(
+    name: String,
+    watch_vars: Vec<Var>,
+    constraint: DelayedConstraintFn<Var, Value>,
+) -> Box<dyn Constraint<Var, Value>>
+where
+    Var: cmp::Eq + PartialEq + Hash + Clone + Debug + 'static,
+    Value: Clone + Debug + 'static,
+{
+    Box::new(Implies {
+        name,
+        watch_vars,
+        constraint: DelayedConstraint::Delayed(constraint),
+    })
+}
+
+pub type DelayedConstraintFn<Var, Value> =
+    Arc<dyn Fn(&[&Value]) -> Box<dyn Constraint<Var, Value>>>;
+
+#[derive(Clone)]
+enum DelayedConstraint<Var, Value> {
+    Delayed(DelayedConstraintFn<Var, Value>),
+    Constraint(Box<dyn Constraint<Var, Value>>),
+}
+
+#[derive(Clone)]
+struct Implies<Var, Value> {
+    name: String,
+    watch_vars: Vec<Var>,
+    constraint: DelayedConstraint<Var, Value>,
+}
+
+impl<Var, Value> Constraint<Var, Value> for Implies<Var, Value>
+where
+    Var: cmp::Eq + PartialEq + Hash + Clone + Debug,
+    Value: Clone + Debug,
+{
+    fn update(
+        &mut self,
+        assignment: &mut Assignment<Var, Value>,
+        key: fn(&Value) -> &str,
+    ) -> Result<bool, ConstraintError<Var, Value>> {
+        let mut updated = false;
+        if let DelayedConstraint::Delayed(delayed) = &self.constraint {
```

Comment:
can this be an exhaustive match?

also, why is the `Constraint` branch unreachable?

Reply:
The first check `if let DelayedConstraint::Delayed ...` converts either exists or converts a delayed constraint to an immediate constraint, so when it reaches `let DelayedConstraint::Constraint ...` the constraint must be an immediate one.

Add a small comment to clarify.